### PR TITLE
[browser][mt] Don't allow relink duplicates of `dotnet.native.worker.js`

### PR DESCRIPTION
**Pull request by ilonatommy**
Don't allow relink duplicates and check for the mt native files in WBT.

### CODE REVIEW

File: src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs [hunk 1]
```diff
@@ -105,7 +105,10 @@ public void AssertBundle(AssertWasmSdkBundleOptions assertOptions)
             _ => throw new ArgumentOutOfRangeException(nameof(assertOptions.ExpectedFileType))
         };
         string buildType = assertOptions.IsPublish ? "publish" : "build";
-        foreach (string nativeFilename in new[] { "dotnet.native.wasm", "dotnet.native.js" })
+        var nativeFilesToCheck = assertOptions.RuntimeType == RuntimeVariant.MultiThreaded ?
+            new[] { "dotnet.native.wasm", "dotnet.native.js", "dotnet.native.worker.js" } :
+            new[] { "dotnet.native.wasm", "dotnet.native.js" };
```
Comment (maraf) [inline src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs:110]:
```suggestion
        var nativeFilesToCheck = new List<string>() { "dotnet.native.wasm", "dotnet.native.js" }
        if (assertOptions.RuntimeType == RuntimeVariant.MultiThreaded)
            nativeFilesToCheck.Add("dotnet.native.worker.js");
```

File: src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs [hunk 2]
```diff
@@ -105,7 +105,10 @@ public void AssertBundle(AssertWasmSdkBundleOptions assertOptions)
             _ => throw new ArgumentOutOfRangeException(nameof(assertOptions.ExpectedFileType))
         };
         string buildType = assertOptions.IsPublish ? "publish" : "build";
-        foreach (string nativeFilename in new[] { "dotnet.native.wasm", "dotnet.native.js" })
+        var nativeFilesToCheck = assertOptions.RuntimeType == RuntimeVariant.MultiThreaded ?
```
Comment (maraf) [inline src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs:108]: Could we assert there is exactly one file in the collections that are supposed to contain one file?
It probably should happen in the base class

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets [hunk 3]
```diff
@@ -225,7 +225,7 @@ Copyright (c) .NET Foundation. All rights reserved.
 
       <!-- Remove dotnet.js/wasm from runtime pack, in favor of the relinked ones in @(WasmNativeAsset) -->
       <ReferenceCopyLocalPaths Remove="@(ReferenceCopyLocalPaths)"
-                               Condition="@(WasmNativeAsset->Count()) > 0 and ( '%(FileName)' == 'dotnet' or '%(FileName)' == 'dotnet.native' ) and ('%(Extension)' == '.wasm' or '%(Extension)' == '.js')" />
+                               Condition="@(WasmNativeAsset->Count()) > 0 and ( '%(FileName)' == 'dotnet' or '%(FileName)' == 'dotnet.native' or '%(FileName)' == 'dotnet.native.worker' ) and ('%(Extension)' == '.wasm' or '%(Extension)' == '.js')" />
```
Comment (maraf) [inline src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:228]: Could we automate this step? 
Maybe remove all `ReferenceCopyLocalPaths` that exist in `WasmNativeAsset`

File: src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs (hunk 2 above)
Comment (ilonatommy) [reply src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs]: You mean that there are not 2 files with same names in the same path? I thought about it but technically it's impossible to have 2 files of same name at same location, so checking for it should not be needed.

Comment (maraf) [reply src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs]: I would like to check content of `blazor.boot.json`

Comment (ilonatommy) [reply src/mono/wasm/Wasm.Build.Tests/WasmSdkBasedProjectProvider.cs]: That's already done in https://github.com/dotnet/runtime/blob/eeadd653e1982d7037a93a9ab38129c07336e7db/src/mono/wasm/Wasm.Build.Tests/ProjectProviderBase.cs#L455
it compares the count of resources in `blazor.boot.json` with fixed expected resources set that contains one `dotnet.native.worker.js` for MT.

File: src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets [hunk 4]
```diff
@@ -220,12 +220,14 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
   <Target Name="_ResolveWasmOutputs" DependsOnTargets="ResolveReferences;PrepareResourceNames;ComputeIntermediateSatelliteAssemblies;_ResolveWasmConfiguration;_WasmNativeForBuild;_GetWasmRuntimePackVersion">
+      <CreateItem Include="@(ReferenceCopyLocalPaths)" Condition="'%(Filename)' != '' and '@(ReferenceCopyLocalPaths)' != '' and '@(WasmNativeAsset)' != ''">
```
[the remark this answers is not in the document]
Comment (maraf) [reply src/mono/nuget/Microsoft.NET.Sdk.WebAssembly.Pack/build/Microsoft.NET.Sdk.WebAssembly.Browser.targets:223]: It's based on strings, but the leading and trailing `;` ensures only whole filename substring is matched

**Comment by ilonatommy** [comment]
Locally sizes of `dotnet.native.worker*.js` differ fine in relink (release): 8kB in runtime pack vs 4kB in publish bundle.

**Comment by maraf** [comment]
This doesn't seem correct

image

**Comment by ilonatommy** [comment]
> Did you run publish?

Yes, the failing test `SimpleMultiThreadedTests.BlazorPublishRunTest` is running publish so I run it was well to reproduce.

**Comment by ilonatommy** [comment]
It got fixed by comparing more metadata than just size. Shouldn't sizes of packages files be deterministic on same SDKs?

`SimpleMultiThreadedTests.BlazorPublishRunTest` WBT locally:
```
AssertFile
/workspaces/runtime/artifacts/bin/dotnet-latest/packs/Microsoft.NETCore.App.Runtime.Mono.multithread.browser-wasm/9.0.0-dev/runtimes/browser-wasm/native/dotnet.native.worker.js and 
/workspaces/runtime/artifacts/bin/Wasm.Build.Tests/Debug/net9.0/linux-x64/wbt artifacts/blazor_mt_Release_yoplqrvn_bzt/bin/Release/net9.0/publish/wwwroot/_framework/dotnet.native.worker.9.0.0-dev.d9rgdagevy.js .
Expected to be the same = False.
And are same = False.
Have same sizes = False.
Size1: 6883
Size2: 2819
```
While relinked size matches with the error reported on CI, the runtime pack file size does not. In both cases `9.0.0-preview.2.24109.4` was used.

**Review by maraf — APPROVED** [review]

**Pull request merged**